panther_analysis_tool/detection.py

```diff
@@ -189,3 +249,597 @@ def get_module(  # pylint: disable=arguments-differ
         self, identifier: str, body: str
     ) -> ModuleType:
         return super().from_string(identifier, body, self._tmp_dir)
+
+
+# pylint: disable=too-many-instance-attributes
+class Detection:
+    """Panther detection metadata and imported module."""
+
+    # pylint: disable=too-many-branches,too-many-statements
+    def __init__(self, config: Mapping, default_type: str):
+        """Create new detection from a dict.
+        Args:
+            config: Dictionary that we expect to have the following keys:
+                analysisType: either RULE, SCHEDULED_RULE, or POLICY
+                id: Unique detection identifier
+                body: The detection body
+                versionId: The version of the detection
+                (Optional) path: The detection module path
+                (Optional) dedupPeriodMinutes: The period during which
+                the events will be deduplicated
+        """
+        self.logger = get_logger()
+
+        # Check for required string fields
+        for each_field in ["id", "versionId"]:
+            if not (each_field in config) or not isinstance(config[each_field], str):
+                raise AssertionError('Field "%s" of type str is required field' % each_field)
+
+        if not (config.get("body") or config.get("path")):
+            raise ValueError('one of "body", "path" must be defined')
+
+        if config.get("body") and config.get("path"):
+            raise ValueError('only one of "body", "path" must be defined')
+
+        if not any((isinstance(config.get("body"), str), isinstance(config.get("path"), str))):
+            raise TypeError('"body", "path" parameters must be of string type')
+
+        # backwards compatible for data from before we returned this
+        self.detection_type = config.get("analysisType", default_type)
```

Comment:
Ah, I think `default_type` should not be a constructor parameter. You can define the fallback here, because it might be confusing when someone sees the policy constructor and the still see `rule` in this parameter.

Reply:
I guess I'm not as familiar with the backwards compatibility reqs here - were there policies that had a defined "rule" function? I removed it from constructor and created a property for the default type (so we can use `policy` as default in policies and `rule` as default in rules)